doc/design/switchdev-refactoring.md

```diff
@@ -0,0 +1,180 @@
+
+### Use Cases
+
+* As a developer I don't want to maintain the code with similar logic in two places:
+switchdev bash scrips and **sriov-network-operator-daemon** code.
+* As a developer and a user I want to have only one set of systemd services that handle
+both _**switchdev**_  configurations and `systemd` mode.
+* As a user I want `configurationMode` to work the same way for NICs with
+_**legacy**_ and _**switchdev**_ configurations.
+* As a user I want to apply _**switchdev**_ configuration for NIC by the **sriov-network-operator-daemon**
+without reboot (in case if reboot is not required by other logic,
+e.g. kernel parameters configuration, FW configuration).
+
+### Goals
+
+* it should be possible to apply  _**switchdev**_ configuration in the **sriov-network-operator-daemon** without reboot.
+* the code used by `daemon` and `systemd` modes to handle _**switchdev**_ and _**legacy**_
+configurations should be the same Golang code.
+* `configurationMode` option should work the same for NICs with _**legacy**_ and _**switchdev**_ configurations.
+* the operator should use unified systemd services which will be deployed only
+if the operator works in the `systemd` mode.
+* `systemd` mode should be changed to support 2 stage configuration: 
+pre system network manager (NetworkManager or netplan) and after system network manager.
+  _Note: This is required to support all use-cases supported by the current switchdev implementation._
+
+
+### Non-Goals
+
+* replace _Externally Manage PF_ feature
+* remove all shell scripts from the code
+
+## Proposal
+
+1. Drop existing bash-based implementation which is used for NICs with _**switchdev**_ configuration
+2. Modify _generic_ and _virtual (if required)_ plugins to support _**switchdev**_ configuration
+3. Modify code related to the _Externally Manage PF_ feature 
+to support _**switchdev**_ configuration
+4. Modify `systemd` mode flow to handle devices with both  _**legacy**_  and _**switchdev**_  configurations
+5. Split `systemd` mode system service to two parts:
+    - `pre` - executes before NetworkManager/netplan and OVS
+    - `after` - executes after NetworkManager/netplan and OVS
+
+```mermaid
+---
+title: Proposed logic
+---
+flowchart TD;
+    operatorMode["SriovOperatorConfig.configurationMode"];
+    inDaemon["apply configurations(legacy and switchdev) in
+     sriov-network-operator-daemon"]
+    systemdSystemd["create `pre` and `after` systemd services(go code)
+    to handle configurations(legacy and switchdev) on boot"]
+
+    operatorMode-- daemon --> inDaemon;
+    operatorMode-- systemd --> systemdSystemd;
+
+```
+
+
+### Workflow Description
+
+Users using only NICs with _**legacy**_ SRIOV configurations will not need to change their workflow.
+The operator should behave for these configurations the same way as it does now.
+
+Users using NICs with _**switchdev**_ configurations will need to explicitly set operator's 
```

Comment:
AFAIU this statement is true also for legacy mode. isn't it?

Reply:
Right, If you want legacy or switchdev configs to apply on boot, you should use systemd mode. 
This sentence here is to highlight that users of switchdev configs will need to change config (change `configurationMode` to systemd) if they want to get the "old" behavior. 
The users of legacy SRIOV configs don't need to do anything to preserve existing behavior.